### PR TITLE
primitives to allow lifting from operations from TorchSharp tensors to DiffSharp tensors

### PR DESCRIPTION
**Pull request by dsyme**
These primitives allow us to lift primal operations implemented in TorchSharp up to DiffSharp when we can assume the user is using the Torch backend 

@pkese you may be intereted in this

The point is to allow the user to program freely with TorchSharp operations to generate primal tensors.

We use the name `primalRawTorch` (like `primalRaw` but getting a `torch.Tensor`) since it discards differentiability.

### CODE REVIEW

**Comment by dsyme** [comment]
@gbaydin I'm finding these necessary for porting some of the Pythong code in the work we're doing.

**Comment by dsyme** [comment]
@gbaydin If it's ok I'll merge this to unblock my porting work? The functions are simple enough that I don't think they need tests.  (We can also adjust the names later)

**Comment by dsyme** [comment]
As an example of code that now ports more straightforwardly, consider
```python
def weight_init_conv2d(out_channels, in_channels, kernel_height, kernel_width):
    w = torch.zeros(out_channels, in_channels, kernel_height, kernel_width)
    nn.init.kaiming_uniform_(w, a=math.sqrt(5))
    return w
```
We want to be able to port this torch code that creates tensors from scratch without much thought, e.g.
```fsharp
let weight_init_conv2d(out_channels: int, in_channels: int, kernel_height: int, kernel_width: int) =
    let w = torch.zeros [| int64 out_channels; int64 in_channels; int64 kernel_height; int64 kernel_width |]
    torch.nn.init.kaiming_uniform_(w, a=Math.Sqrt(5)) |> ignore
    dsharp.ofTorchTensor w
```
This similarly applies to torch input/output and torchvision image manipulation code.

**Comment by gbaydin** [comment]
I'm looking at it now, will comment very soon.

**Comment by gbaydin** [comment]
I think `w.toTensor()` on a torch tensor is not good because w is already a tensor for a torch user.

I think the design should be something like the following. It's ok if these currently don't support differentiation and only move the primal data. This can be mentioned in the docs and that should be good enough.

```fsharp
dsharpTensor = dsharp.fromTorch(torchTensor)
torchTensor = dsharpTensor.toTorch()
```

This is in the same spirit with `torch.from_numpy` that is used to take a numpy ndarray in python and make a torch tensor out of it. I think we shouldn't have anything beyond this simple api.

**Comment by dsyme** [comment]
I'm fine with `dsharp.fromTorch`

For the other one I think there is a fair risk of people (including myself) not realising they're losing differentiability, e.g. say TorchSharp has a tensor function `torch.foo`, then they will do

```fsharp
let foo (x: Tensor) = x.toTorch() |> torch.foo |> dsharp.fromTorch
```
but using that in a model will cause some (likely obscure) error when training, or even worse the differentiation calculation will still succeed under the assumption that the tensor is constant.  Is seems wrong

It seems to me `primalAbc` is currently our consistent terminology to indicate we're losing differentiability and so I'd say stick with that.

**Comment by dsyme** [comment]
Is adding `DiffSharp.Torch` namespace ok?

**Comment by gbaydin** [comment]
> Is adding DiffSharp.Torch namespace ok?

I actually don't like it and would be happy to avoid that if possible.

**Comment by dsyme** [comment]
> I actually don't like it and would be happy to avoid that if possible.

So just add to `DiffSharp`?

**Comment by gbaydin** [comment]
> So just add to DiffSharp?

Yes, if you agree. :)

**Comment by dsyme** [comment]
Done. I changed to `toTorch` as I don't mind.  But be aware it's a trap for people (like converting to numpy is - but it's a bit more subtle as they might be used to the idea that Torch tensors support gradients).

**Comment by gbaydin** [comment]
Thank you!

> Done. I changed to `toTorch` as I don't mind.  But be aware it's a trap for people (like converting to numpy is - but it's a bit more subtle as they might be used to the idea that Torch tensors support gradients). 

I see the point and I agree. If we had some possibility to raise warnings instead of exceptions, this would be a good place to raise a warning. I think F# and dotnet in general doesn't have a mechanism for that if I'm not mistaken. We could print a warning to stderr or something.

**Comment by dsyme** [comment]
>  I think F# and dotnet in general doesn't have a mechanism for that if I'm not mistaken. 

There are various things, including CompilerMessageAttribute.  We should extend that to allow "informational" messages, which are much softer than warnings.

**Comment by gbaydin** [comment]
What PyTorch does for this type of things is to use https://docs.python.org/3/library/warnings.html and keep running without an exception. They frequently use this for things like tensor api that will be deprecated in future versions, warning about problems that can be created by double backward calls in autograd, etc.

**Comment by dsyme** [comment]
> What PyTorch does for this type of things is to use https://docs.python.org/3/library/warnings.html and keep running without an exception. They frequently use this for things like tensor api that will be deprecated in future versions, warning about problems that can be created by double backward calls in autograd, etc.

Yes, in F# you can print warnings, or use `CompilerMessageAttribute`, or the .NET `ObsoleteAttribute`.

**Comment by gbaydin** [comment]
> Yes, in F# you can print warnings, or use CompilerMessageAttribute, or the .NET ObsoleteAttribute.

Awesome, I didn't know!

**Comment by gbaydin** [comment]
This all looks good to me

**Comment by dsyme** [comment]
thanks!
